### PR TITLE
LinkedList & Doubly Linked List

### PR DESCRIPTION
**Pull request by codepcl**
Estimado Dr

Adjunto las siguientes tareas:

CLinkedList: operator<< 
CLinkedList: operator>> 
CLinkedList: print, read 
CLinkedList: add 2 param to insert 
CLinkedList: remove general iterator
CLinkedList: add KeyNode
CDoubleLinkedList: operator<<
CDoubleLinkedList: operator>>
CDoubleLinkedList: print, read
CDoubleLinkedList: add 2 param to insert
CDoubleLinkedList: add KeyNode

image

Estare atento al feedback

Saludos cordiales

### CODE REVIEW

File: linkedlist.h [hunk 1]
```diff
@@ -98,7 +115,7 @@ class LinkedList
   public:
     LinkedList() {}
     // TODO add LinkedValueType value
-    void    insert(value_type &elem) { insert_forward(elem);  }
+    void    insert(value_type key, LinkedValueType value) { insert_forward(key, value);  }
```
Comment (edson2495) [inline linkedlist.h]: OK.

File: linkedlist.h [hunk 2]
```diff
@@ -163,10 +180,31 @@ class LinkedList
         throw "hola excepcion"; // Create custom exception pending
     }
     // TODO add print
+    void print        (ostream &os){
```
Comment (edson2495) [inline linkedlist.h]: It works but you might wanna to do it better. Maybe using your iterator.

File: linkedlist.h [hunk 3]
```diff
@@ -163,10 +180,31 @@ class LinkedList
...
+        os << "{" << pTmp->getData() << ", " << pTmp->getValue() << "}, ";
+        pTmp = pTmp->getpNext();
+      }
+    }
 };
 
 // TODO add operator<<
-
+template <typename T>
+ostream &operator<<(ostream &os, LinkedList<T> &obj){
```
Comment (edson2495) [inline linkedlist.h]: OK.

File: linkedlist.h [hunk 4]
```diff
@@ -163,10 +180,31 @@ class LinkedList
...
+    obj.print(os);
+    return os;
+}
 // TODO add operator>>
+template <typename T>
+istream & operator>>(istream &is, LinkedList<T> &obj){
```
Comment (edson2495) [inline linkedlist.h]: NOT OK. Use test.txt file.

Comment (edson2495) [reply linkedlist.h]: Try to use it.

File: doublelinkedlist.h [hunk 5]
```diff
@@ -1,100 +1,222 @@
...
+    Node *pNew = CreateNode(elem, value);
+    if (m_pTail)
+      m_pTail->setpNext(pNew);
+    m_pTail = pNew;
+    if (!m_pHead)
+      m_pHead = pNew;
+    m_size++;
+  }
+
+  void print(ostream &os)
```
Comment (edson2495) [inline doublelinkedlist.h:267]: Make it better like the comment in linkedlist.h

File: doublelinkedlist.h [hunk 6]
```diff
@@ -1,100 +1,222 @@
...
     }
+    os << endl;
+    return os;
+  }
 };
 
+template <typename T>
+inline ostream &operator<<(ostream &os, DoubleLinkedList<T> &obj)
+{
+  obj.print(os);
```
Comment (edson2495) [inline doublelinkedlist.h:305]: OK.

File: doublelinkedlist.h [hunk 7]
```diff
@@ -1,100 +1,222 @@
...
+  return os;
+}
+
+template <typename T>
+inline istream &operator>>(istream &is, DoubleLinkedList<T> &obj)
+{
+  obj.insert(obj.value_type, obj.LinkedValueType);
```
Comment (edson2495) [inline doublelinkedlist.h]: NOT using test.txt file.
Try to use it in demo.cpp.

File: doublelinkedlist.h [hunk 8]
```diff
@@ -1,100 +1,222 @@
...
-  public:
-    backward_iterator(Container *pContainer, Node *pNode) : Parent (pContainer,pNode) {}
-    backward_iterator(myself &other)  : Parent (other) {}
-    backward_iterator(myself &&other) : Parent(other) {}
-
-  public:
-    backward_iterator operator++() { Parent::m_pNode = ((Node *)Parent::m_pNode)->getpPrev();  
-                                     return *this;
-                                   }
+class backward_iterator : public general_iterator<Container, class backward_iterator<Container>> //
```
Comment (edson2495) [inline doublelinkedlist.h]: Remove inheritance.

**Comment by edson2495** [comment]
No estas eliminando herencia en los iteradores (forward y backward).

**Comment by edson2495** [comment]
Sugiero cada cosa que implementas y que va a ser evaluada mandar un screenshot de su uso.

**Comment by codepcl** [comment]
image

**Review by edson2495 — APPROVED** [review]

**Comment by edson2495** [comment]
Everything looks fine, just make your print function better.

**Pull request closed without merging**